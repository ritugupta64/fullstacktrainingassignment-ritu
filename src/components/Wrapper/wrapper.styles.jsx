import { css } from 'styled-components'

const Styles = css`

display:flex;
flex-wrap: wrap;
padding: 1rem;
.container__left-column{
        flex:  0 0 20%;
        padding-right: 3rem;
     } 
.container__right-column{
       flex: 0 0 80%;
    }
 .container__top-container{
   display: flex;
}  
@media only screen and (max-width: 767px){
   .container__left-column{
      flex-basis: 100%;
      padding-right: 0;
   }
   .container__right-column{
      flex-basis: 100%;
   } 
} 

@media only screen and (min-width: 768px) and (max-width:991px){
   .container__right-column{
      flex-basis: 75%;
   } 
} 

}`

export default Styles
