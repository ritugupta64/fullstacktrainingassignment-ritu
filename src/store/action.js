export const getCharacterData = (data, dispatch) => {
   dispatch({
       type: 'getCharacter',
       payload: data
   })
}