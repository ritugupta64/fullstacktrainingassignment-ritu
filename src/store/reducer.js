
const initialState = {
    character: []
}
 
const Reducer = (state, action) => {
    switch(action.type){
        case 'getCharacter':
        return{
            ...state,
            character:action.payload
        }
        default:
            return state || initialState
    }

}

export {
    initialState, Reducer
}