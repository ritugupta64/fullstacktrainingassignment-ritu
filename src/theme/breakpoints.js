export default {
  sm: '768px',
  xs: '480',
};
