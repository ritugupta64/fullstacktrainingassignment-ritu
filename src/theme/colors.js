export default {
  secondary: '#976117',
  whiteColor: '#fff',
  blackColor: '#000',
  paraColor: '#5D5D5D',
  greyColor: '#333333',
  bgGrey: '#808080',
  fontGrey: '#1F232A',
};
