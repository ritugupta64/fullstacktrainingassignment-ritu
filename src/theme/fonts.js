const base = 16;

export default {
  fontSize: num => `${Number(num / base)}rem`,
};
