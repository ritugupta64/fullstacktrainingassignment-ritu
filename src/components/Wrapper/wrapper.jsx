import React, { useState, useEffect, useContext } from 'react'
import styled from 'styled-components'

import { GlobalContext } from '../../store/context'
import { getCharacterData } from '../../store/action'
import CharacterList from '../organism/Characters/CharacterList/characterList'
import { fetchWrapper } from '../../utilities'
import Heading from '../atoms/Heading'
import Select from '../atoms/Select'
import SearchBox from '../molecules/SearchBox/search'
import FilterList from '../organism/FilterList/filterList'
import Tags from '../molecules/Tags'
import styles from './wrapper.styles'
import Loader from '../atoms/loader'

const Speicies = ['Human', 'Mythology', 'Other Species']
const Gender = ['Male', 'Female']
const Origin = [
  'Unknown',
  'Post-Apocalyptic Earth',
  'Nuptia 4',
  'Other Origins',
]

const selectByID = ['Ascending', 'Descending']

const Base_url = 'https://rickandmortyapi.com/api/character/'

const Wrapper = ({ className }) => {
  const { dispatch } = useContext(GlobalContext)
  const [data, setData] = useState([])
  const [isLoading, setIsLoading] = useState()
  const [sort, setSort] = useState('Ascending')
  const [search, setSearch] = useState('')
  const [filters, setFilters] = useState({})

  const handleSorting = (e) => setSort(e.target.value)
  const sortedData = (content) => {
    let sortedContent = { ...content }
    if (sort === 'Descending') {
      sortedContent = sortedContent?.results?.sort((a, b) => b.id - a.id)
    } else {
      sortedContent = sortedContent?.results?.sort((a, b) => a.id - b.id)
    }

    return { ...content, results: sortedContent }
  }

  const handleSearchValue = (value) => setSearch(value)

  const handleFilters = (checkedValues, removedValues, id) => {
    const filtersObj = { ...filters }
    if (removedValues) {
      filtersObj[id] = filtersObj[id]?.filter((itm) => itm !== removedValues)
    } else {
      filtersObj[id] = filtersObj[id]
        ? [...filtersObj[id], checkedValues]
        : [checkedValues]
    }

    setFilters(filtersObj)
  }

  const handleCloseFilter = (name, category) => {
    handleFilters(null, name, category)
    document.getElementsByName(name)[0].checked = false
  }

  const fetchUrl = async () => {
    const searchfilters = `?${search ? `&name=${search}` : ''}`
    const filtersString = Object.keys(filters)
      .map((itm) => `&${filters[itm].map((ftr) => `${itm}=${ftr}`).join('&')}`)
      .join('&')

    const url = await fetchWrapper(
      `${Base_url}${searchfilters}${filtersString}`
    )
    setData(url)
    setIsLoading(true)
    sortedData(url)
    getCharacterData(url, dispatch)
  }
  useEffect(() => {
    fetchUrl()
  }, [search, filters])

  useEffect(() => {
    setData(sortedData(data))
  }, [sort])

  return !isLoading ? (
    <Loader>Loading...</Loader>
  ) : (
    <div className={className}>
      <aside className="container__left-column">
        <Heading>Filters</Heading>

        <FilterList
          filterList={Speicies}
          title="Species"
          getSelectedFilter={(amend, remove) =>
            handleFilters(amend, remove, 'species')
          }
        />
        <FilterList
          filterList={Gender}
          title="Gender"
          getSelectedFilter={(amend, remove) =>
            handleFilters(amend, remove, 'gender')
          }
        />
        <FilterList
          filterList={Origin}
          title="Origin"
          getSelectedFilter={(amend, remove) =>
            handleFilters(amend, remove, 'origin')
          }
        />
      </aside>
      <section className="container__right-column">
        <Tags
          tags={Object.keys(filters)?.reduce((accumulator, currentValue) => {
            const tagsCategory = filters[currentValue].map((itm) => ({
              category: currentValue,
              name: itm,
            }))
            return [...accumulator, ...tagsCategory]
          }, [])}
          getCloseItem={(category, name) => handleCloseFilter(name, category)}
          className="tags"
        />
        <div className="container__top-container">
          <SearchBox
            getValue={handleSearchValue}
            labelText="Search by Name"
            className="search"
          />

          <Select selectList={selectByID} SelectOption={handleSorting} />
        </div>

        <CharacterList characterList={data?.results} />
      </section>
    </div>
  )
}

export default styled(Wrapper)`
  ${styles};
`
