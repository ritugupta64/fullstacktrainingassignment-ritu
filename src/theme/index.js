import colors from './colors';
import breakpoints from './breakpoints';
import tokens from './tokens'

export default {
  colors,
  breakpoints,
  tokens
};
