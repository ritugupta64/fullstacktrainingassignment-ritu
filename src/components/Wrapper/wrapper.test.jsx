import React from 'react'
import { render, cleanup } from '@testing-library/react'
import { ThemeProvider } from 'styled-components'
import { GlobalContextProvider } from '../../store/context'
import theme from '../../theme'

import Wrapper from './wrapper'

let result
const dispatch = jest.fn()

const initialState = {
  character: [],
}

afterEach(cleanup)

it('renders', () => {
  const { asFragment } = render(
    <ThemeProvider theme={theme}>
      <GlobalContextProvider data={{ state: initialState, dispatch }}>
        <Wrapper />
      </GlobalContextProvider>
    </ThemeProvider>
  )
  expect(asFragment()).toMatchSnapshot()
})
