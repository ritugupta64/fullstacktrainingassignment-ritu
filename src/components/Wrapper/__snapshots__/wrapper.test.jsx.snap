// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render a button should match snapshot 1`] = `undefined`;

exports[`renders 1`] = `
<DocumentFragment>
  <div
    class="sc-AxheI cWRoBT"
  >
    Loading...
  </div>
</DocumentFragment>
`;
